File: src/copy-trading.ts
import { CopyTrading, COPY_TRADING_REST_API_PROD_URL } from '@binance/copy-trading';
import inquirer from 'inquirer';
import {
    decodeSelectedEntities,
    getConfigurationRestAPI,
    getUserAgent,
    isEmpty,
    readStdinObj,
    getParsedArgs,
} from './utils';

const parsedArgs = getParsedArgs();
const isFullDescription = parsedArgs?.['full-description'] ?? false;

const getClient = () => {
    if (!process.env?.LOG_LEVEL) {
        process.env.LOG_LEVEL = 'ERROR';
    }
    process.env.BINANCE_CONNECTOR_JS_USER_AGENT = getUserAgent('copy-trading');

    let basePath = COPY_TRADING_REST_API_PROD_URL;

    const configurationRestAPI = getConfigurationRestAPI(parsedArgs?.profile, 'copy-trading');

    if (process.env.BINANCE_COPY_TRADING_BASE_PATH) {
        basePath = process.env.BINANCE_COPY_TRADING_BASE_PATH;
    } else if (configurationRestAPI && configurationRestAPI['basePath']) {
        basePath = configurationRestAPI['basePath'];
    }

    let client;
    let hasConfig = false;
    if (configurationRestAPI !== null) {
        hasConfig = true;
        client = new CopyTrading({
            configurationRestAPI: { ...configurationRestAPI, basePath },
        });
    } else {
        client = new CopyTrading({
            configurationRestAPI: {
                apiKey: '',
                basePath,
            },
        });
    }

    return { client, hasConfig };
};

const copyTradingCommands: any[] = [];

copyTradingCommands.push({
    command: 'get-futures-lead-trader-status',
    describe:
        'Authentication required. ' +
        decodeSelectedEntities(
            `Get Futures Lead Trader Status

Weight: 20`,
            isFullDescription
        ),
    builder: (yargsCmd: any) => {
        return yargsCmd.options({
            'recv-window': {
                describe: decodeSelectedEntities(''),
                type: 'string',
                group: 'Command Options:',
            },
            json: {
                describe: 'Send all fields as JSON',
                type: 'string',
                group: 'JSON Options:',
            },
        });
    },
    handler: async (options: any) => {
        const questions: any = [];
        const stdinObj: any = readStdinObj();

        if (!isEmpty(stdinObj)) {
            options = { ...options, ...stdinObj };
        }

        if (options.json) {
            options = { ...options, ...JSON.parse(options.json) };
            delete options.json;
        }

        const { client, hasConfig } = getClient();
        if (!hasConfig) {
            console.error(
                'get-futures-lead-trader-status is signed. Please create a profile using `binance-cli profile create`.'
            );
            process.exitCode = 1;
            return;
        }

        if (questions.length > 0) {
            const answers = await inquirer.prompt(questions);
            options = { ...options, ...answers };
        }

        try {
            const response = await client.restAPI.getFuturesLeadTraderStatus(options);
            const responseData = await response.data();
            console.log(JSON.stringify(responseData, null, 2));
        } catch (e: any) {
            console.log(e.message);
            return;
        }
    },
});

copyTradingCommands.push({
    command: 'get-futures-lead-trading-symbol-whitelist',
    describe:
        'Authentication required. ' +
        decodeSelectedEntities(
            `Get Futures Lead Trading Symbol Whitelist

Weight: 20`,
            isFullDescription
        ),
    builder: (yargsCmd: any) => {
        return yargsCmd.options({
            'recv-window': {
                describe: decodeSelectedEntities(''),
                type: 'string',
                group: 'Command Options:',
            },
            json: {
                describe: 'Send all fields as JSON',
                type: 'string',
                group: 'JSON Options:',
            },
        });
    },
    handler: async (options: any) => {
        const questions: any = [];
        const stdinObj: any = readStdinObj();

        if (!isEmpty(stdinObj)) {
            options = { ...options, ...stdinObj };
        }

        if (options.json) {
            options = { ...options, ...JSON.parse(options.json) };
            delete options.json;
        }

        const { client, hasConfig } = getClient();
        if (!hasConfig) {
            console.error(
                'get-futures-lead-trading-symbol-whitelist is signed. Please create a profile using `binance-cli profile create`.'
            );
            process.exitCode = 1;
            return;
        }

        if (questions.length > 0) {
            const answers = await inquirer.prompt(questions);
            options = { ...options, ...answers };
        }

        try {
            const response = await client.restAPI.getFuturesLeadTradingSymbolWhitelist(options);
            const responseData = await response.data();
            console.log(JSON.stringify(responseData, null, 2));
        } catch (e: any) {
            console.log(e.message);
            return;
        }
    },
});

export default {
    command: 'copy-trading',
    description: 'Binance Copy Trading REST API',
    builder: (yargs: any) => {
        yargs.usage('Usage: binance-cli copy-trading <command> [options]');
        copyTradingCommands.forEach((command: any) => {
            yargs.command(command);
        });
    },
};

File: src/copy_trading.rs
use crate::utils::{
    decode_selected_entities, get_client_configuration, init_user_agent, read_json_as,
    read_stdin_as,
};
use binance_sdk::config::{ConfigurationRestApi, PrivateKey};
use binance_sdk::constants::COPY_TRADING_REST_API_PROD_URL;
use binance_sdk::copy_trading::CopyTradingRestApi;
use binance_sdk::copy_trading::rest_api::*;
use clap::{Args, Subcommand};
use std::env;
use std::io;
use std::io::{Error, ErrorKind};

fn get_client(profile: Option<&str>, is_signed: bool) -> Result<RestApi, Error> {
    init_user_agent("copy-trading");

    let client_config = get_client_configuration(profile, "copy-trading").unwrap();
    let api_env = env::var("BINANCE_API_ENV")
        .ok()
        .or(client_config.env)
        .unwrap_or_else(|| "prod".to_string());

    let base_path = client_config.base_path.unwrap_or(match api_env.as_str() {
        "prod" => COPY_TRADING_REST_API_PROD_URL.to_string(),
        _ => {
            return Err(Error::new(
                ErrorKind::InvalidInput,
                "Invalid api env, valid values: prod",
            ));
        }
    });

    let mut builder = ConfigurationRestApi::builder().base_path(base_path);

    if is_signed {
        builder = builder
            .api_key(client_config.api_key)
            .api_secret(client_config.api_secret);

        if client_config.private_key.is_some() {
            builder = builder.private_key(PrivateKey::File(client_config.private_key.unwrap()));
        }
    }

    let rest_conf = builder
        .build()
        .map_err(|err| Error::new(ErrorKind::Other, err))?;

    Ok(CopyTradingRestApi::from_config(rest_conf))
}

#[derive(Args, Debug)]
struct GetFuturesLeadTraderStatusArgs {
    #[arg(help = r#"Request validity window in milliseconds"#, long)]
    recv_window: Option<i64>,
    #[arg(short = 'i', long)]
    interactive: bool,
    #[arg(help = r#"Send all fields as JSON"#, long)]
    json: Option<String>,
    #[arg(help = r#"Select a profile"#, long)]
    profile: Option<String>,
}
#[derive(Args, Debug)]
struct GetFuturesLeadTradingSymbolWhitelistArgs {
    #[arg(help = r#"Request validity window in milliseconds"#, long)]
    recv_window: Option<i64>,
    #[arg(short = 'i', long)]
    interactive: bool,
    #[arg(help = r#"Send all fields as JSON"#, long)]
    json: Option<String>,
    #[arg(help = r#"Select a profile"#, long)]
    profile: Option<String>,
}

#[derive(Subcommand)]
pub enum CopyTradingCommands {
    #[command(
        about = decode_selected_entities(r#"Get Futures Lead Trader Status

Weight(IP): 1

Security Type: TRADE"#, false),
    )]
    GetFuturesLeadTraderStatus(GetFuturesLeadTraderStatusArgs),
    #[command(
        about = decode_selected_entities(r#"Get Futures Lead Trading Symbol Whitelist

Weight(IP): 1

Security Type: USER_DATA"#, false),
    )]
    GetFuturesLeadTradingSymbolWhitelist(GetFuturesLeadTradingSymbolWhitelistArgs),
}

pub async fn handle_copy_trading_command(command: CopyTradingCommands) -> anyhow::Result<()> {
    match command {
        CopyTradingCommands::GetFuturesLeadTraderStatus(args) => {
            get_futures_lead_trader_status(args).await
        }

        CopyTradingCommands::GetFuturesLeadTradingSymbolWhitelist(args) => {
            get_futures_lead_trading_symbol_whitelist(args).await
        }
    }
}

async fn get_futures_lead_trader_status(
    args: GetFuturesLeadTraderStatusArgs,
) -> anyhow::Result<()> {
    let rest_client = get_client(args.profile.as_deref(), true)?;

    let params = match read_stdin_as::<GetFuturesLeadTraderStatusParams>() {
        Some(params) => params,
        None => match args.json {
            Some(json) => {
                read_json_as::<GetFuturesLeadTraderStatusParams>(json).ok_or_else(|| {
                    io::Error::new(io::ErrorKind::InvalidInput, "failed to parse json params")
                })?
            }
            None => GetFuturesLeadTraderStatusParams::builder()
                .recv_window(args.recv_window)
                .build()?,
        },
    };

    // Make the API call
    let response = rest_client.get_futures_lead_trader_status(params).await?;

    let data = response.data().await?;
    println!("{}", serde_json::to_string_pretty(&data).unwrap());
    Ok(())
}

async fn get_futures_lead_trading_symbol_whitelist(
    args: GetFuturesLeadTradingSymbolWhitelistArgs,
) -> anyhow::Result<()> {
    let rest_client = get_client(args.profile.as_deref(), true)?;

    let params = match read_stdin_as::<GetFuturesLeadTradingSymbolWhitelistParams>() {
        Some(params) => params,
        None => match args.json {
            Some(json) => read_json_as::<GetFuturesLeadTradingSymbolWhitelistParams>(json)
                .ok_or_else(|| {
                    io::Error::new(io::ErrorKind::InvalidInput, "failed to parse json params")
                })?,
            None => GetFuturesLeadTradingSymbolWhitelistParams::builder()
                .recv_window(args.recv_window)
                .build()?,
        },
    };

    // Make the API call
    let response = rest_client
        .get_futures_lead_trading_symbol_whitelist(params)
        .await?;

    let data = response.data().await?;
    println!("{}", serde_json::to_string_pretty(&data).unwrap());
    Ok(())
}
